### PR TITLE
fix: lenient gate comment detection to prevent duplicates from other tools (#451)

### LINKED ISSUE #451: Bug: upsert-gate-review-comment creates duplicate when a valid gate comment already exists from another source
## Summary

On PR #450, two `pre_approval_gate` comments were posted 72 seconds apart for the same head SHA (`e284c2e`). The first (id 4615274563) was a detailed review comment. The second (id 4615284620) was a concise version. Both had all 5 contract fields present.

## Root cause

The `upsert-gate-review-comment.mjs` helper searches for existing gate comments by matching the gate name + head SHA in the comment body. If a gate comment was posted via a different tool (e.g. direct `gh pr comment`), the upsert helper may fail to find it and create a duplicate.

## Fix

Make the upsert helper more tolerant of existing gate comments posted by other tools:
- Parse ALL PR issue comments for gate markers, not just those posted by this helper
- If a valid gate comment (matching gate name + head SHA) already exists, update it instead of creating a new one
- Only create a new comment if no existing gate+head match is found

## Acceptance criteria

- [x] Upsert helper detects gate comments posted via any tool
- [x] No duplicate gate comments for same gate + head SHA
- [x] npm test passes

## Source

Observed on PR #450. Related to audit #441, Scan A finding #8.

### PR DESCRIPTION
**Pull request by mfittko**
## Change Summary

Fix duplicate gate comment detection when gate comments are posted via other tools. Make `parseGateReviewCommentFields` more lenient by adding a fallback regex search for gate name + head SHA anywhere in the comment body, not just in structured field format.

## Scope / Context

The `upsert-gate-review-comment.mjs` helper searches for existing gate comments by matching gate name + head SHA. When a gate comment was posted via a different tool (e.g. direct `gh pr comment`) without the exact structured format, the parser failed to detect it, causing a duplicate comment.

**Fix**: `parseGateReviewCommentFields` now includes a lenient fallback that scans the entire comment body for:
- Gate name (`draft_gate` or `pre_approval_gate`) as a whole word
- A 7-64 character hex SHA

This allows `parseGateReviewCommentMarkerBody` (used by `summarizeGateReviewCommentMarkers`) to detect gate comments regardless of format.

## Acceptance Criteria

- [x] Marker parser detects gate comments posted in non-standard format
- [x] No duplicate gate comments for same gate + head SHA  
- [x] Structured format comments still parsed correctly (no regression)
- [x] `npm test` passes (1725 tests)

## Definition of Done

- [x] Lenient fallback added to `parseGateReviewCommentFields`
- [x] 9 new tests: lenient format detection, structured format unchanged, null when no match
- [x] All existing tests pass (839 core, 886 scripts)

## Non-goals

- No changes to `parseGateReviewCommentBody` strictness (still requires all fields)
- No new API calls
- No changes to upsert-gate-review-comment CLI interface

Closes #451

### CODE REVIEW

**Comment by mfittko** [comment]
### Gate review: `draft_gate`

**Reviewed head SHA:** `21b97f693076e32918451d8295579b74e550fb03`
**Verdict:** clean

**Findings summary:** All tests pass (1725 total). 9 new tests for lenient gate comment parsing. No existing behavior changed.

**Next action:** Ready for Copilot review. Run: gh pr ready 453

File: packages/core/src/github/copilot-helpers.mjs [hunk 1]
```diff
@@ -114,6 +114,26 @@ function parseGateReviewCommentFields(body) {
...
+      }
+    }
+
+    if (!fields.headSha) {
+      const shaMatch = flatBody.match(/\b([0-9a-f]{7,64})\b/iu);
+      if (shaMatch) {
+        fields.headSha = normalizeGateReviewHeadSha(shaMatch[1]);
+      }
+    }
+  }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/github/copilot-helpers.mjs:152]: Fixed in 99a0963: strip GitHub #issuecomment-NNN URLs from body before lenient SHA matching. Added test confirming URL with comment ID before actual head SHA still resolves correctly.

File: packages/core/test/copilot-helpers.test.mjs [hunk 2]
```diff
@@ -178,3 +178,109 @@ test("summarizeCopilotReviews ignores non-Copilot reviews", () => {
...
+// ── Lenient gate comment parsing (#451) ───────────────────────────────────
+
+test("parseGateReviewCommentMarkerBody detects gate+head in non-standard format", () => {
+  const body = "pre_approval_gate check for head e284c2e341: all clear!";
+  const result = parseGateReviewCommentMarkerBody(body);
+  assert.notEqual(result, null);
+  assert.equal(result.gate, "pre_approval_gate");
+  assert.equal(result.headSha, "e284c2e341");
+  assert.equal(result.contractComplete, false); // no verdict/next-action fields
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/copilot-helpers.test.mjs:191]: Fixed in 99a0963: added test 'parseGateReviewCommentMarkerBody lenient SHA ignores github comment URLs' covering this exact edge case.

**Review by mfittko — COMMENTED** [review]
Re-requesting Copilot review for commit 99a0963 (fixed URL stripping before SHA detection).

File: packages/core/src/github/copilot-helpers.mjs [hunk 3]
```diff
@@ -114,6 +114,31 @@ function parseGateReviewCommentFields(body) {
...
+        fields.gate = normalizeGateReviewName(gateMatch[1]);
+      }
+    }
+
+    if (!fields.headSha) {
+      const shaMatch = flatBody.match(/\b([0-9a-f]{7,64})\b/iu);
+      if (shaMatch) {
+        fields.headSha = normalizeGateReviewHeadSha(shaMatch[1]);
+      }
+    }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/github/copilot-helpers.mjs:151]: Fixed in 55434d9: two-stage SHA detection now prefers (head|sha|commit) context markers first, falling back to any hex token after URL stripping.

File: packages/core/test/copilot-helpers.test.mjs [hunk 4]
```diff
@@ -178,3 +178,121 @@ test("summarizeCopilotReviews ignores non-Copilot reviews", () => {
...
+  // e.g. URL contains #issuecomment-4615274563 which has 10 decimal digits
+  // that would match [0-9a-f]{7,64} — ensure we strip URLs first
+  const body = "pre_approval_gate for head e284c2e341: all clear!\n\n" +
+    "See https://github.com/mfittko/pi-dev-loops/pull/450#issuecomment-4615274563 for details.";
+  const result = parseGateReviewCommentMarkerBody(body);
+  assert.notEqual(result, null);
+  assert.equal(result.gate, "pre_approval_gate");
+  assert.equal(result.headSha, "e284c2e341");
+  assert.equal(result.contractComplete, false);
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/copilot-helpers.test.mjs:297]: Fixed in 55434d9: added test 'parseGateReviewCommentMarkerBody lenient SHA ignores plain-text numeric ID before head SHA' covering this exact case.

File: packages/core/src/github/copilot-helpers.mjs [hunk 5]
```diff
@@ -114,6 +114,40 @@ function parseGateReviewCommentFields(body) {
...
+
+    if (!fields.headSha) {
+      // Prefer SHA following a "head" context marker to avoid false
+      // matches on plain-text numeric IDs (issue/comment IDs, etc.)
+      // Example: "pre_approval_gate for head e284c2e341" or "commit abc1234def"
+      const ctxShaMatch = flatBody.match(
+        /(?:head|sha|commit)\s*(?:sha)?\s*[:=]?\s*`?\b([0-9a-f]{7,64})\b`?/iu
+      );
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/github/copilot-helpers.mjs:138]: Fixed in e41c466: \\b added around context keywords; GATE_REVIEW_NAMES used for canonical gate list; test added for 'ahead 4615274563, head e284c2e341' case.

File: packages/core/src/github/copilot-helpers.mjs [hunk 6]
```diff
@@ -114,6 +114,40 @@ function parseGateReviewCommentFields(body) {
...
+  // Lenient fallback: detect gate name and head SHA anywhere in body
+  // Handles comments posted via other tools without structured field format
+  if (!fields.gate || !fields.headSha) {
+    const flatBody = body.replace(/\*\*/gu, "").replace(/`/gu, "");
+
+    if (!fields.gate) {
+      const gateMatch = flatBody.match(/\b(draft_gate|pre_approval_gate)\b/iu);
+      if (gateMatch) {
+        fields.gate = normalizeGateReviewName(gateMatch[1]);
+      }
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/src/github/copilot-helpers.mjs:129]: Fixed in e41c466: \\b added around context keywords; GATE_REVIEW_NAMES used for canonical gate list; test added for 'ahead 4615274563, head e284c2e341' case.

File: packages/core/test/copilot-helpers.test.mjs [hunk 7]
```diff
@@ -178,3 +178,131 @@ test("summarizeCopilotReviews ignores non-Copilot reviews", () => {
...
+test("parseGateReviewCommentMarkerBody lenient SHA ignores plain-text numeric ID before head SHA", () => {
+  // Plain-text "comment 4615274563" is 10 decimal digits that would match [0-9a-f]{7,64}
+  // Context-based parser picks SHA after "head", not the numeric ID
+  const body = "pre_approval_gate: comment 4615274563 for head e284c2e341 all clear!";
+  const result = parseGateReviewCommentMarkerBody(body);
+  assert.notEqual(result, null);
+  assert.equal(result.gate, "pre_approval_gate");
+  assert.equal(result.headSha, "e284c2e341");
+  assert.equal(result.contractComplete, false);
+});
```
[the remark this answers is not in the document]
Comment (mfittko) [reply packages/core/test/copilot-helpers.test.mjs:308]: Fixed in e41c466: \\b added around context keywords; GATE_REVIEW_NAMES used for canonical gate list; test added for 'ahead 4615274563, head e284c2e341' case.

**Comment by mfittko** [comment]
### Gate review: `pre_approval_gate`

**Reviewed head SHA:** `e41c4662572ef2a13a17ba1034d0954a9b741ae8`
**Verdict:** clean

**Findings summary:** All 1728 tests pass. 3 rounds of Copilot review: addressed URL stripping, context-based SHA detection, word boundaries, canonical gate names. 12 lenient parsing tests covering edge cases (URL numeric IDs, plain-text IDs, ahead substring, head/sha/commit context). CI green.

**Next action:** Ready to merge. Merge authorized.
